### DISCUSSION TITLE
search functionality

### DISCUSSION THREAD

**Discussion by sanketgarade** [mukta-strot/marathi-shabd#25] [Ideas]
@the-kaustubh  I had some points to discuss, based on which we can decide what all can/cannot be done and plan those.

- [x] make the search site styling (colours, font etc.) consistent with the home page
- [x] if answer to #10 is that backend database is possible, then how it can be developed, utilised and what will be its benefit etc.? also is this solution possible within the confines of github pages or we need to link to some external entity?
- [x] do you have any other ideas about if we should have any more features to the search functionality that can be useful?
- [x] is there a way to take user input on the website for the missing words and add them to some other csv/database file where we can read those and then add them to our main database file?
- [x] also do you have any other suggestions/ideas in general about the project?

**Comment by the-kaustubh** [comment]
- Yes, we can style the page differently. The current styling was kept very plain and simple according to some of the sites that you referenced (e.g. the recipe site). So, yes, styling could be made better, but I'll need some more inputs from you to proceed.

- About database: As the documentation says, we can only host static site, which means github cannot host a database. Although it is possible to have online databases such as [Mongo Atlas](https://www.mongodb.com/cloud/atlas). They have some free tiers and some of them are paid ones. **Benefit of adding database**: What is the procedure of changing a certain word in db.csv? we are required to fork the repo, make changes, and then open a PR, all this just to change a word. By adding a database, we could add or modify words and examples from the hosted site.

- **Ideas for other functionalities**: As of now, since we have tags field in the csv file, I think it is underutilised. There could be a search by tags options, or suggestion for similar words based on tags, etc.
- So yes, we **can** take user input for missing words or meanings or even examples. But, the changes can't be updated into db.csv, as it requires us to follow github contribution procedure. One of the solution is to have a database as explained earlier.

As of now, these are my only suggestions.

**Reply by sanketgarade** [reply]
Thanks. 

1. **Styling** - let me know what inputs you need from me. 
2. **Database** - ok the free one at Mongo Atlas has 5 GB storage which I think will be ok for us right? If so, then you can start a POC at your end on a separate branch for this, as per your convenience. We may need to create an account there I think. Should I create one? 
3. **Searching using tags etc.** - yes we can do that but this we can keep for later I think. 
4. **Users being able to fill in missing words** is a "must have" feature I think. So once the database thing is setup we can go for this. 


Let's also create separate issues for all these so that we can track them individually.

**Reply by the-kaustubh** [reply]
1. Ya so, I would like to know what exactly is to be done in the styling aspect.
2. Yes, please create one account for Atlas, also we need to think about hosting for backend of our site, something like heroku would suffice. They have free tier as well.
3. Fine by me. Anything else you think is doable at this point, add an issue about it someone might work on it. I will too if I can.
4. Yes, so the prerequisite is to have a backend which I will add in couple of days, till then we need to decide about hosting and database. I can take care of heroku if you say so.

**Reply by sanketgarade** [reply]
Ok I'll create separate issues for all these later today when I get time. 

A just had a question regarding the database and backend etc. 
Will the database and backend data/code be hidden behind a login wall? And accessible to only 1 person (account holder)?
Because my target is to have everything open. 


1. If so, I was wondering if we could look at open alternatives for all this, where the stuff is visible to all, or at least visible once you create an account and have the access to the database and backend.


2. I was also thinking that if we could use the external database website exclusively for below purpose
  - a place to save the main database file of words and meaning etc. 
  - to add new words to the main database
  - And a separate database where end-users can add missing words and information (via a submission form on the website), which we can move to the main database later on review. 

3. And have a copy of this external database in our repo in csv format and use that for the browse and search functions on the website. This will remove dependence on external website for the core functionality. 


What do you think?

**Reply by the-kaustubh** [reply]
So the login or no login is upto us. We should keep searching open for all.

1. Later we could think about ranking system based on review or upvotes for meaning of the words and use-in-sentence example. Though this is very much in the future.
2. Yes you are absolutely right, that's what we are going to implement more or less.
3. Although this part will be quite slow, using csv for querying for words is not better than using a database made specifically for these purposes. So I think database should be used for search functions. As reading from a csv is not particularly efficient approach.

So in short, if we are going to let anyone add words/meanings/examples then we could think of adding logins, and rank the better meanings ia upvotes or likes, but that's in the future. For now, I'll complete the search part and implement it using backend and frontend, instead of reading from csv directly. In addition to all the python scripts we have, we could also add one to convert csv file in db query file and the other way around.

**Reply by sanketgarade** [reply]
1. Ok
2. Ok
3. Ok let's go for database. By "efficiency" do you mean time efficiency? So for the end user the search will work faster if we use a database instead of csv?


> In addition to all the python scripts we have, we could also add one to convert csv file in db query file and the other way around.

Yes let's have these scripts to convert the database file from its native format to csv and also one to do the opposite. (That's what you meant, right?)

**Reply by sanketgarade** [reply]
For manjo atlas and heroku, I'll create accounts today and let you know the details via mail. So that you can take it further from there.
